File: examples/swift/flappy_bird/Sources/FlappyBird/Bird.swift
// Bird.swift

import GoudEngine

final class Bird {
    private let game: GoudGame
    private let movement: Movement
    private let animator: BirdAnimator

    private(set) var x: Float
    private(set) var y: Float

    static let width: Float = 34
    static let height: Float = 24

    init(game: GoudGame) {
        self.game = game
        movement = Movement(gravity: GameConstants.gravity, jumpStrength: GameConstants.jumpStrength)
        x = Float(GameConstants.screenWidth) / 4
        y = Float(GameConstants.screenHeight) / 2
        animator = BirdAnimator(game: game, x: x, y: y)
    }

    func initialize() {
        animator.initialize()
    }

    func reset() {
        x = Float(GameConstants.screenWidth) / 4
        y = Float(GameConstants.screenHeight) / 2
        movement.velocity = 0
        animator.reset()
    }

    func update(deltaTime: Float) -> Bool {
        var didFlap = false

        if game.isKeyPressed(key: .SPACE) || game.isMouseButtonPressed(button: .LEFT) {
            didFlap = movement.tryJump(deltaTime: deltaTime)
        }

        movement.applyGravity(deltaTime: deltaTime)
        var yPosition = y
        movement.updatePosition(positionY: &yPosition, deltaTime: deltaTime)
        y = yPosition

        animator.update(deltaTime: deltaTime, x: x, y: y, rotation: movement.rotation)

        return didFlap
    }

    func draw() {
        animator.draw()
    }
}

File: examples/android/flappy_bird/app/src/main/java/com/goudengine/flappybird/Bird.kt
package com.goudengine.flappybird

import com.goudengine.core.GoudGame
import com.goudengine.input.GamepadButton
import com.goudengine.input.Key
import com.goudengine.types.Color

/**
 * The player-controlled bird.
 *
 * Supports three input methods for flapping:
 * - Touch: tap anywhere on the screen (primary mobile input)
 * - Keyboard: press Space (useful in Android emulator)
 * - Gamepad: A button / South face button
 */
class Bird(game: GoudGame) {
    var x: Float = GameConstants.SCREEN_WIDTH / 3f
    var y: Float = GameConstants.SCREEN_HEIGHT / 2f
    private var velocity: Float = 0f
    private var jumpCooldown: Float = 0f
    private val animator = BirdAnimator(game)

    fun update(game: GoudGame, dt: Float) {
        jumpCooldown -= dt
        val scaledDt = dt * GameConstants.TARGET_FPS

        if (shouldFlap(game) && jumpCooldown <= 0f) {
            velocity = GameConstants.JUMP_STRENGTH * GameConstants.TARGET_FPS
            jumpCooldown = GameConstants.JUMP_COOLDOWN
        }

        velocity += GameConstants.GRAVITY * scaledDt
        y += velocity * dt

        animator.update(dt)
    }

    fun render(game: GoudGame) {
        val texture = animator.currentFrame()
        game.drawSprite(
            texture, x, y,
            GameConstants.BIRD_WIDTH.toFloat(),
            GameConstants.BIRD_HEIGHT.toFloat(),
            0f, Color.white()
        )
    }

    fun reset() {
        x = GameConstants.SCREEN_WIDTH / 3f
        y = GameConstants.SCREEN_HEIGHT / 2f
        velocity = 0f
        jumpCooldown = 0f
    }

    /**
     * Returns true if the player wants to flap via any supported input method.
     *
     * Input methods:
     * 1. Touch - tap anywhere (touchId 0 = primary finger)
     * 2. Keyboard - Space key (works in Android emulator)
     * 3. Gamepad - A/South button on gamepad 0
     */
    private fun shouldFlap(game: GoudGame): Boolean {
        // Touch input: primary mobile input method
        if (game.isTouchJustPressed(0)) return true

        // Keyboard input: useful for Android emulator testing
        if (game.isKeyJustPressed(Key.Space)) return true

        // Gamepad input: A/South button on gamepad 0.
        if (game.isGamepadButtonJustPressed(0, GamepadButton.South.value)) return true

        return false
    }
}
